### PR TITLE
Encoding and transaction creation overhaul and tests

### PR DESCRIPTION
**Pull request by dexX7**
Also work in progress. :)

Will ping, once this is final. But feel free to drop comments of course, if there is something to note.

### CODE REVIEW

**Comment by zathras-crypto** [comment]
Fantastic, thanks for this mate :)

**Comment by dexX7** [comment]
Hey @zathras-crypto, please check out [7a62a01/src/omnicore_encoding.cpp](https://github.com/dexX7/bitcoin/blob/7a62a011091e83b79566aba590f79e69acceafed/src/omnicore_encoding.cpp).

Is this too much, or too little documentation? It's a bit difficult for me to find a middle ground.

**Comment by zathras-crypto** [comment]
> @zathras-crypto, please check out 7a62a01/src/omnicore_encoding.cpp.

Fine by me mate :)

**Comment by dexX7** [comment]
@zathras-crypto: the memo fields for granting and revoking tokens are optional, which strictly means there should not even be a null character, in case the memo field is empty. But are we still able to differentiate between "junk" and "memo", if no null character is required? How does the decoding/interpretation logic currently handle this case?

**Comment by zathras-crypto** [comment]
IIRC the memo field isn't even parsed by OmniCore on decoding currently.  I'll have to double check.

But yeah, I don't see why we couldn't test to see if the memo parameter was empty and drop the null terminator if it's not used.

**Comment by dexX7** [comment]
> But yeah, I don't see why we couldn't test to see if the memo parameter was empty and drop the null terminator if it's not used.

But would it still be possible to identify memo messages properly? Consider the following cases to handle:
1. [version: 0] [type: 55] [property: 8] [amount: 1000] (no terminiator)
2. [version: 0] [type: 55] [property: 8] [amount: 1000] [0x00]
3. [version: 0] [type: 55] [property: 8] [amount: 1000] ["thanks for the tokens, bob!" 0x00]
4. [version: 0] [type: 55] [property: 8] [amount: 1000] [some random junk 0x00]

Case 1: must not invalidate the transaction, this is very important
Case 2: should be fine, because it's simply an empty message
Case 3: should be fine, because it's a null terminated message
Case 4: I guess the random junk would be interpreted as memo, which is okay

Case 1 is the most important one, and if that's followed, then the rest doesn't really matter, as it doesn't affect consensus.

**Comment by dexX7** [comment]
Worth to note that the decoding logic is sort of mean, when it comes to decoding of text fields and the length of 255 characters. See my email: "Maybe consensus issue: string fields" (2015-02-14)

**Comment by zathras-crypto** [comment]
You should find it stops at case 1 full stop at the current time because neither `logicMath_GrantTokens()` or `logicMath_RevokeTokens()` reads any additional data from the packet (where the memo field would be).

`CMPTransaction::step2_Value()` only takes us up to the property and amount bytes - anything after that is ignored unless the packet is read in later steps (which isn't the case for grants and revokes). 

TL:DR; if we wanted to add support for parsing and displaying memos in grants and revokes we need to definitely consider this - but for now there shouldn't be a consensus concern.

**Comment by zathras-crypto** [comment]
> See my email: "Maybe consensus issue: string fields" (2015-02-14)

Sure, let me pull that up and take a look :)

**Comment by dexX7** [comment]
Great. :)

**Comment by zathras-crypto** [comment]
Interesting - hmm - so what options do we have in this scenario:
1) Truncate
2) Invalidate message
3) ???

Definitely agree a behaviour should be defined.

**Comment by dexX7** [comment]
On the interpretation side, only the first 255 characters are taken, so this is covered.

But I sort of think this violates the spec, and I'm wondering, if similar transactions should be invalidated. At least it should be reflected in the spec, in one way or the other.

**Comment by dexX7** [comment]
Second time in a row where the Windows x64 build hit the 50 minutes limit.. :S FWIW: you should be able to trigger a rebuild for a single job manually.

**Comment by dexX7** [comment]
I tend to say: go ahead, this can be merged.

Not, because I don't think there can be more tests, but to resolve the merge conflict, and to see how everything fits together. And I have to make up my mind, to come up with more useful tests, but I'm sort of out of ideas - at the very moment at least. :) It would probably make sense to continue with RPC/bitcoin-spock tests, in particular more tests for the DEx and crowdsales, to cover a broader range.

**Comment by zathras-crypto** [comment]
Reran, successful, merging - thanks again bud :)
